### ISSUE TITLE
vmuser takes long time(approx 30 mins) to become operational

### ISSUE THREAD

**Issue by uprashan** [VictoriaMetrics/VictoriaMetrics#7721]
### Is your question request related to a specific component?

victoriametrics cluster, vmoperator, vmauth

### Describe the question in detail

Benchmarking with 10K vmusers created on a single cluster and using vmoperator to sync and vmauth as proxy for authorization. 
We see a behaviour where although a vmuser gets created successfully. It takes unusual amount of time for the newly created vmuser to become operational. The more the number of users, the more time it takes for it become operational and start accepting metrics for the tenant.

How can we address this issue?

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/troubleshooting/
- [X] vmagent - https://docs.victoriametrics.com/vmagent/#troubleshooting
- [X] vmalert - https://docs.victoriametrics.com/vmalert/#troubleshooting

**Comment by uprashan** [comment]
In addition, vmauth gets into crashloop/OOM..

`    Command:
      /bin/prometheus-config-reloader
    Args:
      --reload-url=http://localhost:8427/-/reload
      --config-envsubst-file=/opt/vmauth/config.yaml
      --config-file=/opt/vmauth-config-gz/config.yaml.gz
    State:       Waiting
      Reason:    CrashLoopBackOff
    Last State:  Terminated
      Reason:    OOMKilled`

**Comment by Haleygo** [comment]
First, you need to increase resources for vmauth to prevent OOM. Then, check the resource usage of ETCD, the API server, and vm-operator to ensure there’s no throttling. 
The sync process may be slow for several reasons:
1. The ETCD&API server is slow to return all vmuser resources to vm-operator;
2. vm-operator requires time&resources to generate config for all 10k vmusers for each vmuser create&update operation(including duplicate checks, etc.)
3. vmauth needs time&resources to reload new config.

Is 10K vmusers a typical scenario for you in the future, or is it just for benchmarking? 
For better performance, I recommend keeping vmusers under 1K.

**Comment by uprashan** [comment]
Hi @Haleygo . Thank you. No, it isn't just for benchmarking. But, we do have use cases that could include upto 10K and more users.

**Comment by Haleygo** [comment]
Will all of them be active(have requests all the time)? 
What tool will you use to manage these users, or are you migrating them from a third-party platform?

**Comment by uprashan** [comment]
Yes, we expect all of them to be active and have incoming requests. The VMUser will be created through a custom API call as and when a new user/account is created in our system. We are building the architecture to ensure it can scale to 10K or more users in the future.
What would you recommend if we need to handle 10K or more users in the cluster?

**Comment by Haleygo** [comment]
I see the following issues:
1. Like you obeserved, vmauth requires more resources to handle the load. Also, vmauth exposes metrics per user, such as:
```
vmauth_user_requests_total
vmauth_user_request_backend_errors_total
vmauth_user_request_duration_seconds
```
With 10k users, single vmauth instance exposes over 50k metrics, which can be challenging to collect for vmagent or prometheus.
2. Having 10k vmuser(or any other resources) in a single K8s cluster bring significant pressure on both Kubernetes components(etcd&API server) and vm-operator. Each create&update&delete operation on vmuser results in large request to Kubernetes, requiring the vm-operator to regenerate and verify the configuration.

The second issue is non-trivial to solve for now. 
If your users are infrequently updated and the majority of requests is adding users, I'd recommend to not use vm-operator initially. Instead, using a script in the pipeline to update vmauth config file directly and see if vmauth functions as expected.

**Comment by uprashan** [comment]
2. The only operation we would carry out is to create users.
   With update of vmauth config to add new user, this would mean that we need to reload the config causing brief downtime for other users during the config reload? Is the understanding correct?

**Comment by Haleygo** [comment]
You can still use `vmauth` in vm-operator, which creates vmauth pod with a config-reloader sidecar, it calls the vmauth `/-/reload` endpoint when it sees((default `watch-interval=3m`, you can decrease it if needed)) change under `--config-file(mount your vmauth config here)`.

**Comment by f41gh7** [comment]
Hello, currently operator performance is mostly limited by:
- allocated CPUs
- Kubernetes client configuration.


 v0.52.0 release of operator raises default requests limit from 5 to 50 requests per second. And it should reduce processing time of VMUsers creations to ~1 minute for your scale. 

 I did a benchmark and with new default value for 5_000 VMusers it requires 4 CPU limit burst and 0.2 CPU  requests to maintain a local cache for the given amount of objects. 

 Related commit https://github.com/VictoriaMetrics/operator/commit/100b8df932e6d826199f060b5b5eb785b03bf062

**Comment by Haleygo** [comment]
Hi @uprashan ,
Do you have a chance to run new benchmarks on the latest version? Is this issue still relevant?
